### PR TITLE
Refactor to use Game Loop and Structs.

### PR DESCRIPTION
**Pull request by danielemdv**

### CODE REVIEW

File: sketch/sketch.ino [hunk 1]
```diff
@@ -1,363 +1,486 @@
...
 
-  while (true) {
 
-    // Flash currently highlighted player
-    highlightFlashOn = doPlayerSelection_flashHighlightedPlayer(highlightFlashOn);
+// START ====== GLOBAL STUFF ====================
 
-    // MISSING: Show highlighted player triggers and chamber size on some sort of display.
+#define PLAYER_COUNT 4
+byte startingPlayer = -1;
```
Comment (danielemdv) [inline sketch/sketch.ino]: Need to clean a bunch of stuff like this.

File: sketch/sketch.ino [hunk 2]
```diff
@@ -1,363 +1,486 @@
...
+  int state;
+  int pin;
+  int playerIndex;
+  unsigned long prevBlinkMillis;
+
+  // Default constructor
+  PlayerLed() : state(-1), pin(-1), playerIndex(-1), prevBlinkMillis(0) { }
+  
+  // Custom constructor
+  PlayerLed(int pinIn, int playerIndexIn) {
```
Comment (relegarum) [inline sketch/sketch.ino:120]: You also can use the 
: state(HIGH),
pin(pinIn),
....
{
}

File: sketch/sketch.ino [hunk 3]
```diff
@@ -1,363 +1,486 @@
...
   }
-  pinMode(buzzerPin, OUTPUT);
-  pinMode(greenButtonPin, INPUT_PULLUP);
-  pinMode(redButtonPin, INPUT_PULLUP);
-  pinMode(knobPin, INPUT);
-}
 
-void loop() {
+  // Copy assignment operator
+  Player& operator=(const Player& other) {
```
Comment (relegarum) [inline sketch/sketch.ino:68]: More like a move.
Player& operator=(Player&& other) {
... 
}

File: sketch/sketch.ino [hunk 4]
```diff
@@ -1,363 +1,486 @@
...
-    } else {
-      if (digitalRead(redButtonPin) == LOW) {
-        doPlayerSelection_updateHighlightedPlayer(1); // Select clockwise active player
-        Serial.println("Player selection updated via button");
-      }
-    }
+// LEDs
+const int playerLedPins[] = {2,3,4,5}; // Need to add more pins if more players wanted.
+PlayerLed* playerLeds = new PlayerLed[PLAYER_COUNT];
+const int ledBlinkDuration = 300;
```
Comment (relegarum) [inline sketch/sketch.ino]: static constexpr int ledBlinkDuration = 300;

File: sketch/sketch.ino [hunk 5]
```diff
@@ -1,363 +1,486 @@
...
-    // Note: Probably more elegant to change state and handle animation/sounds in new state.
+// Buzzer
+const int buzzerPin = 6;
+struct Buzzer buzzer = {buzzerPin, 0};
 
-    delay(100); //  This delay value important and can affect selection knob accuracy.
-  }
+//  Button
+const int buttonAPin = 7;
+const int buttonBPin = 8;
```
Comment (relegarum) [inline sketch/sketch.ino]: All constansts prefer to use constexpr

File: sketch/sketch.ino [hunk 6]
```diff
@@ -1,363 +1,486 @@
...
+// Selection input type
+enum SelectionMode: int {
+  Knob,
+  Button
+};
+
+SelectionMode currentSelectionMode = Button; // Change this depending on secondary input type
+
+// GameState
+enum GameState: int {
```
Comment (relegarum) [inline sketch/sketch.ino]: enum class GameState: int {
...
};
Will prevent putting `int` where you want GameState.

File: sketch/sketch.ino (hunk 3 above)
Comment (relegarum) [reply sketch/sketch.ino:68]: nit: check if you are assign against the same obj
```cpp 
if (this == &other)
{
return *this;
} 
```

File: sketch/sketch.ino [hunk 7]
```diff
@@ -1,363 +1,486 @@
...
+const int ledIntervalDuration = 500;
 
-    if (digitalRead(greenButtonPin) == LOW) {
-      break;
-    }
-    // Read shoot button press, call logic and then exit loop if pressed
-    // Note: Probably more elegant to change state and handle animation/sounds in new state.
+// Buzzer
+const int buzzerPin = 6;
+struct Buzzer buzzer = {buzzerPin, 0};
```
Comment (relegarum) [inline sketch/sketch.ino:273]: struct Buzzer buzzer = {buzzerPin, 0};

could be just Buzzer buzzer = ...;

Comment (danielemdv) [reply sketch/sketch.ino:273]: Yeah, this one was to disappear a warning in my IDE :P

File: sketch/sketch.ino (hunk 5 above)
Comment (danielemdv) [reply sketch/sketch.ino]: This is a new thing for me, what’s the difference?

File: sketch/sketch.ino (hunk 3 above)
Comment (danielemdv) [reply sketch/sketch.ino:68]: > More like a move. Player& operator=(Player&& other) { ... }

Oh god, here come the symbols!

Comment (danielemdv) [reply sketch/sketch.ino:68]: Adding the check for the same instance, won't do the move change yet.

File: sketch/sketch.ino (hunk 2 above)
Comment (danielemdv) [reply sketch/sketch.ino:120]: Thanks, not a fan of the syntax, I'll get used to it I guess

File: sketch/sketch.ino (hunk 5 above)
Comment (danielemdv) [reply sketch/sketch.ino]: Sounds pretty cool actually, powerful stuff

**Pull request merged**